Sources/HTTPTypes/HTTPParsedFields.swift

```diff
@@ -0,0 +1,212 @@
+
+struct HTTPParsedFields {
+    private var method: ISOLatin1String?
+    private var scheme: ISOLatin1String?
+    private var authority: ISOLatin1String?
+    private var path: ISOLatin1String?
+    private var extendedConnectProtocol: ISOLatin1String?
+    private var status: ISOLatin1String?
+    private var fields: HTTPFields = .init()
+
+    enum ParsingError: Error {
+        case invalidName
+        case invalidPseudoName
+        case invalidPseudoValue
+        case multiplePseudo
+        case pseudoNotFirst
+
+        case requestWithoutMethod
+        case invalidMethod
+        case requestWithResponsePseudo
+
+        case responseWithoutStatus
+        case invalidStatus
+        case responseWithRequestPseudo
+
+        case trailersWithPseudo
+
+        case multipleContentLength
+        case multipleContentDisposition
+        case multipleLocation
+    }
+
+    mutating func add(field: HTTPField) throws {
+        if field.name.isPseudo {
+            if !self.fields.isEmpty {
+                throw ParsingError.pseudoNotFirst
+            }
+            switch field.name {
+            case .method:
+                if self.method != nil {
+                    throw ParsingError.multiplePseudo
+                }
+                self.method = field.rawValue
+            case .scheme:
+                if self.scheme != nil {
+                    throw ParsingError.multiplePseudo
+                }
+                self.scheme = field.rawValue
+            case .authority:
+                if self.authority != nil {
+                    throw ParsingError.multiplePseudo
+                }
+                self.authority = field.rawValue
+            case .path:
+                if self.path != nil {
+                    throw ParsingError.multiplePseudo
+                }
+                self.path = field.rawValue
+            case .protocol:
+                if self.extendedConnectProtocol != nil {
+                    throw ParsingError.multiplePseudo
+                }
+                self.extendedConnectProtocol = field.rawValue
+            case .status:
+                if self.status != nil {
+                    throw ParsingError.multiplePseudo
+                }
+                self.status = field.rawValue
+            default:
+                throw ParsingError.invalidPseudoName
+            }
+        } else {
+            self.fields.append(field)
+        }
+    }
+
+    private func validateFields() throws {
+        guard self.fields[values: .contentLength].allElementsSame else {
+            throw ParsingError.multipleContentLength
+        }
+        guard self.fields[values: .contentDisposition].allElementsSame else {
+            throw ParsingError.multipleContentDisposition
+        }
+        guard self.fields[values: .location].allElementsSame else {
+            throw ParsingError.multipleLocation
+        }
+    }
+
+    var request: HTTPRequest {
```

Comment:
NIT: We often write these as `init`s on the types and do the logic in the init instead of making them "conversion" vars. This follows how Swift defines type conversions

Reply:
The ultimate APIs are inits, but I thought they are better as computed properties here so we can keep all the state private.